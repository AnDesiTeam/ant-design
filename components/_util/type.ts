// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#predefined-conditional-types
export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
