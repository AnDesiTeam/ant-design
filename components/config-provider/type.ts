import { ValidateMessages } from 'rc-field-form/lib/interface';
import { RenderEmptyHandler } from './renderEmpty';
import { CSPConfig, DirectionType } from './context';
import { RequiredMark } from '../form/Form';
import { Locale } from '../locale-provider';
import { SizeType } from './SizeContext';

export interface ConfigProviderProps {
  getTargetContainer?: () => HTMLElement;
  getPopupContainer?: (triggerNode?: HTMLElement) => HTMLElement;
  prefixCls?: string;
  iconPrefixCls?: string;
  children?: React.ReactNode;
  renderEmpty?: RenderEmptyHandler;
  csp?: CSPConfig;
  autoInsertSpaceInButton?: boolean;
  form?: {
    validateMessages?: ValidateMessages;
    requiredMark?: RequiredMark;
  };
  input?: {
    autoComplete?: string;
  };
  locale?: Locale;
  pageHeader?: {
    ghost: boolean;
  };
  componentSize?: SizeType;
  direction?: DirectionType;
  space?: {
    size?: SizeType | number;
  };
  virtual?: boolean;
  dropdownMatchSelectWidth?: boolean;
}
