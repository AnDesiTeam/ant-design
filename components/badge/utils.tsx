import { PresetColorTypes } from '../_util/colors';
import type { PresetColorType } from '../_util/colors';

// eslint-disable-next-line import/prefer-default-export
export function isPresetColor(color?: string): boolean {
  return PresetColorTypes.includes(color as PresetColorType);
}
