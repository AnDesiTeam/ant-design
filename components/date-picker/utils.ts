import * as moment from 'moment';

export function formatDate(
  value: moment.Moment | undefined | null,
  format: string | string[] | ((moment: moment.Moment | undefined) => string),
): string {
  if (!value) {
    return '';
  }
  if (Array.isArray(format)) {
    format = format[0];
  }

  if (typeof format === 'function') {
    format = format(value);
  }

  return value.format(format);
}
