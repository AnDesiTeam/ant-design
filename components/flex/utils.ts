import classNames from 'classnames';

import type { FlexProps } from './interface';

export const flexDirectionValues = ['row', 'row-reverse', 'column', 'column-reverse'] as const;

export const flexWrapValues = ['wrap', 'nowrap', 'wrap-reverse'] as const;

export const justifyContentValues = [
  'flex-start',
  'flex-end',
  'start',
  'end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
  'stretch',
  'normal',
  'left',
  'right',
] as const;

export const alignItemsValues = [
  'center',
  'start',
  'end',
  'flex-start',
  'flex-end',
  'self-start',
  'self-end',
  'baseline',
  'normal',
  'stretch',
] as const;

const genClsWrap = (prefixCls: string, wrap: FlexProps['wrap']) => {
  const wrapCls: Record<PropertyKey, boolean> = {};
  flexWrapValues.forEach((cssKey) => {
    wrapCls[`${prefixCls}-wrap-${cssKey}`] = wrap === cssKey;
  });
  return wrapCls;
};

const genClsAlign = (prefixCls: string, align: FlexProps['align']) => {
  const alignCls: Record<PropertyKey, boolean> = {};
  alignItemsValues.forEach((cssKey) => {
    alignCls[`${prefixCls}-align-${cssKey}`] = align === cssKey;
  });
  return alignCls;
};

const genClsJustify = (prefixCls: string, justify: FlexProps['justify']) => {
  const justifyCls: Record<PropertyKey, boolean> = {};
  justifyContentValues.forEach((cssKey) => {
    justifyCls[`${prefixCls}-justify-${cssKey}`] = justify === cssKey;
  });
  return justifyCls;
};

const genClsDirection = (prefixCls: string, direction: FlexProps['direction']) => {
  const directionCls: Record<PropertyKey, boolean> = {};
  flexDirectionValues.forEach((cssKey) => {
    directionCls[`${prefixCls}-direction-${cssKey}`] = direction === cssKey;
  });
  return directionCls;
};

const genMergedAlignCls = (
  prefixCls: string,
  align: FlexProps['align'],
  direction: FlexProps['direction'],
) => {
  if (align) {
    return genClsAlign(prefixCls, align);
  }
  return {
    ...genClsAlign(prefixCls, align),
    [`${prefixCls}-align-stretch`]: direction === 'column' || direction === 'column-reverse',
  };
};

const createFlexClassNames = (
  prefixCls: string,
  props: Pick<FlexProps, 'direction' | 'wrap' | 'justify' | 'align'>,
) => {
  const { direction, wrap, justify, align } = props;
  return classNames({
    ...genClsWrap(prefixCls, wrap),
    ...genClsJustify(prefixCls, justify),
    ...genClsDirection(prefixCls, direction),
    ...genMergedAlignCls(prefixCls, align, direction),
  });
};

export default createFlexClassNames;
