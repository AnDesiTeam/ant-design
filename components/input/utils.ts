import type { ReactNode } from 'react';
import type { InputProps } from './Input';

// eslint-disable-next-line import/prefer-default-export
export function hasPrefixSuffix(props: {
  prefix?: ReactNode;
  suffix?: ReactNode;
  allowClear?: InputProps['allowClear'];
}) {
  return !!(props.prefix || props.suffix || props.allowClear);
}
