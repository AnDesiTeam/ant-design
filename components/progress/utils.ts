import { presetPrimaryColors } from '@ant-design/colors';
import type { CircleProps } from './Circle';
import type { ProgressProps } from './progress';
import warning from '../_util/warning';

export function validProgress(progress?: number) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    warning(
      false,
      'Progress',
      '`success.progress` is deprecated. Please use `success.percent` instead.',
    );
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export const getPercentage = ({ percent, success, successPercent }: ProgressProps) => {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
};

export const getStrokeColor = ({ success = {}, strokeColor }: Partial<CircleProps>): (
  | string
  | Record<PropertyKey, string>
)[] => {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
};

export const getSize = (
  size: ProgressProps['size'],
  type: ProgressProps['type'] | 'step',
  extra?: {
    steps?: number;
  },
): [number, number] => {
  let width: number = -1;
  let height: number = -1;
  if (type === 'step') {
    const steps = extra!.steps!;
    if (typeof size === 'string' || typeof size === 'undefined') {
      width = size === 'small' ? 2 : 14;
      height = 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = 14, height = 8] = size;
    }
    width *= steps;
  } else if (type === 'line') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      height = size === 'small' ? 6 : 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = -1, height = 8] = size;
    }
  } else if (type === 'circle' || type === 'dashboard') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      [width, height] = size === 'small' ? [60, 60] : [120, 120];
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      width = size[0] ?? size[1] ?? 120;
      height = size[0] ?? size[1] ?? 120;
    }
  }
  return [width, height];
};
