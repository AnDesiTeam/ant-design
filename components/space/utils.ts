import type { SpaceSize } from '.';
import type { SizeType } from '../config-provider/SizeContext';

export const isPresetSize = (size: SpaceSize): size is SizeType => {
  if (!size) {
    return false;
  }
  return typeof size === 'string' && ['small', 'middle', 'large'].includes(size);
};

export const isValidNumber = (size: SpaceSize): size is number => {
  if (!size) {
    // 这里故意排除 size = 0 的情况，因为 gap 属性在 CSS 中的默认值就是 0，所以如果用户传0进来的话，我们可以直接忽略
    return false;
  }
  return typeof size === 'number' && !Number.isNaN(size);
};
