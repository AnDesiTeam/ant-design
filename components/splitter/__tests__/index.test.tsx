/* eslint-disable react/no-array-index-key */
import React from 'react';
import { Splitter } from 'antd';
import type { GetProps } from 'antd';

import { fireEvent, render, waitFakeTimer } from '../../../tests/utils';

type PanelProps = GetProps<typeof Splitter.Panel>;

// jsdom 不执行任何布局，因此不会计算值 https://github.com/jsdom/jsdom/issues/1590
global.HTMLElement.prototype.getBoundingClientRect = () =>
  ({
    height: 400,
    width: 400,
  }) as DOMRect;

const SplitterDemo = ({ items = [{}, {}] }: { items?: PanelProps[] }) => (
  <Splitter style={{ width: 400, height: 400 }}>
    {items?.map((item, idx) =>
      item ? (
        <Splitter.Panel key={idx} {...item}>
          {idx}
        </Splitter.Panel>
      ) : (
        <Splitter.Panel key={idx}>{idx}</Splitter.Panel>
      ),
    )}
  </Splitter>
);

describe('Splitter', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  it('should correct render', () => {
    const { container } = render(<SplitterDemo />);
    expect(container?.querySelector('.ant-splitter')).toBeTruthy();
    expect(container?.querySelectorAll('.ant-splitter-panel')?.length).toBe(2);
    expect(container?.querySelector('.ant-splitter-bar')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('should correct render panel size', async () => {
    const { container } = render(
      <SplitterDemo
        items={[{ defaultSize: 20 }, { defaultSize: '45%' }, { defaultSize: '39.4px' }, {}]}
      />,
    );

    await waitFakeTimer();

    expect(container?.querySelectorAll('.ant-splitter-panel')?.[0]).toHaveStyle(
      'flex-basis: calc(20% - 1.5px)',
    );
    expect(container?.querySelectorAll('.ant-splitter-panel')?.[1]).toHaveStyle(
      'flex-basis: calc(45% - 1.5px)',
    );
    expect(container?.querySelectorAll('.ant-splitter-panel')?.[2]).toHaveStyle(
      'flex-basis: calc(10% - 1.5px)',
    );
    expect(container?.querySelectorAll('.ant-splitter-panel')?.[3]).toHaveStyle(
      'flex-basis: calc(25% - 1.5px)',
    );
  });

  it('should resizable', () => {
    const { container } = render(
      <SplitterDemo items={[{ defaultSize: 20 }, { resizable: false }, {}]} />,
    );

    expect(container?.querySelectorAll('.ant-splitter-bar-resizable')?.length).toBe(1);
  });

  it('should collapsible', () => {
    const { container, rerender } = render(
      <SplitterDemo items={[{ defaultSize: 20, collapsible: true }, {}]} />,
    );

    expect(container?.querySelectorAll('.ant-splitter-bar-collapse-icon')?.length).toBe(2);
    expect(container?.querySelector('.ant-splitter-bar-collapse-previous')).toBeTruthy();
    expect(container?.querySelector('.ant-splitter-bar-collapse-next')).toBeTruthy();

    // support collapsible is object
    rerender(
      <SplitterDemo
        items={[
          {
            defaultSize: 20,
            collapsible: {
              prev: true,
            },
          },
          {
            collapsible: {
              next: true,
            },
          },
          {},
        ]}
      />,
    );
    const barNodes = container?.querySelectorAll('.ant-splitter-bar');

    expect(barNodes?.[0]?.querySelector('.ant-splitter-bar-collapse-previous')).toBeTruthy();
    expect(barNodes?.[0]?.querySelector('.ant-splitter-bar-collapse-next')).toBeFalsy();

    expect(barNodes?.[1]?.querySelector('.ant-splitter-bar-collapse-previous')).toBeFalsy();
    expect(barNodes?.[1]?.querySelector('.ant-splitter-bar-collapse-next')).toBeTruthy();
  });

  it('should collapsible click work', () => {
    // previous click
    const { container, rerender } = render(
      <SplitterDemo
        items={[
          {
            size: 20,
            collapsible: {
              prev: true,
            },
          },
          {},
        ]}
      />,
    );
    fireEvent.click(container?.querySelector('.ant-splitter-bar-collapse-previous')!);
    expect(container?.querySelector('.ant-splitter-bar-collapse-previous')).toBeFalsy();
    expect(container?.querySelector('.ant-splitter-bar-collapse-next')).toBeTruthy();

    // next click
    rerender(
      <SplitterDemo
        items={[
          {
            size: 60,
            collapsible: {
              next: true,
            },
          },
          {},
        ]}
      />,
    );
    fireEvent.click(container?.querySelector('.ant-splitter-bar-collapse-next')!);
    expect(container?.querySelector('.ant-splitter-bar-collapse-previous')).toBeTruthy();
    expect(container?.querySelector('.ant-splitter-bar-collapse-next')).toBeFalsy();

    // collapsible is boolean
    rerender(
      <SplitterDemo
        items={[
          {
            size: 10,
            collapsible: true,
          },
          {},
        ]}
      />,
    );
    fireEvent.click(container?.querySelector('.ant-splitter-bar-collapse-next')!);
    expect(container?.querySelector('.ant-splitter-bar-collapse-previous')).toBeTruthy();
    expect(container?.querySelector('.ant-splitter-bar-collapse-next')).toBeFalsy();

    fireEvent.click(container?.querySelector('.ant-splitter-bar-collapse-previous')!);
    expect(container?.querySelector('.ant-splitter-bar-collapse-previous')).toBeTruthy();
    expect(container?.querySelector('.ant-splitter-bar-collapse-next')).toBeTruthy();

    fireEvent.click(container?.querySelector('.ant-splitter-bar-collapse-previous')!);
    expect(container?.querySelector('.ant-splitter-bar-collapse-previous')).toBeFalsy();
    expect(container?.querySelector('.ant-splitter-bar-collapse-next')).toBeTruthy();
  });
});
