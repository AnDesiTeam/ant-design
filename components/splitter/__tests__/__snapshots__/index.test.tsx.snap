// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splitter Nested combinations should work fine. 1`] = `
<div>
  <div
    class="ant-splitter ant-splitter-vertical"
    style="height: 300px; border-radius: 4px; box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);"
  >
    <div
      class="ant-splitter-panel"
      style="flex-basis: 50%;"
    >
      <div
        class="ant-splitter ant-splitter-horizontal"
      >
        <div
          class="ant-splitter-panel"
          style="flex-basis: 33.333333333333336%;"
        >
          <div>
            111
          </div>
        </div>
        <div
          class="ant-splitter-bar"
        >
          <div
            class="ant-splitter-bar-bg"
          />
          <div
            class="ant-splitter-bar-area"
          />
          <div
            class="ant-splitter-bar-resizable"
          />
        </div>
        <div
          class="ant-splitter-panel"
          style="flex-basis: 33.333333333333336%;"
        >
          <div
            class="ant-splitter ant-splitter-vertical"
          >
            <div
              class="ant-splitter-panel"
              style="flex-basis: 50%;"
            >
              <div>
                111
              </div>
            </div>
            <div
              class="ant-splitter-bar"
            >
              <div
                class="ant-splitter-bar-bg"
              />
              <div
                class="ant-splitter-bar-area"
              />
              <div
                class="ant-splitter-bar-resizable"
              />
            </div>
            <div
              class="ant-splitter-panel"
              style="flex-basis: 50%;"
            >
              <div>
                111
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-splitter-bar"
        >
          <div
            class="ant-splitter-bar-bg"
          />
          <div
            class="ant-splitter-bar-area"
          />
          <div
            class="ant-splitter-bar-resizable"
          />
        </div>
        <div
          class="ant-splitter-panel"
          style="flex-basis: 33.333333333333336%;"
        >
          <div>
            2222
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-splitter-bar"
    >
      <div
        class="ant-splitter-bar-bg"
      />
      <div
        class="ant-splitter-bar-area"
      />
      <div
        class="ant-splitter-bar-resizable"
      />
    </div>
    <div
      class="ant-splitter-panel"
      style="flex-basis: 50%;"
    >
      <div
        class="ant-splitter ant-splitter-horizontal"
      >
        <div
          class="ant-splitter-panel wanpan-panel"
          style="background: red; flex-basis: 10%;"
        >
          0
        </div>
        <div
          class="ant-splitter-bar"
        >
          <div
            class="ant-splitter-bar-bg"
          />
          <div
            class="ant-splitter-bar-area"
          />
          <div
            class="ant-splitter-bar-resizable"
          />
        </div>
        <div
          class="ant-splitter-panel"
          style="flex-basis: 45%;"
        >
          1
        </div>
        <div
          class="ant-splitter-bar"
        >
          <div
            class="ant-splitter-bar-bg"
          />
          <div
            class="ant-splitter-bar-area"
          />
          <div
            class="ant-splitter-bar-resizable"
          />
        </div>
        <div
          class="ant-splitter-panel"
          style="flex-basis: 45%;"
        >
          2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter The mousemove should work fine. 1`] = `
<div>
  <div
    class="ant-splitter ant-splitter-horizontal"
  >
    <div
      class="ant-splitter-panel"
      style="flex-basis: 50%;"
    >
      <div
        class="ant-splitter ant-splitter-horizontal"
      >
        <div
          class="ant-splitter-panel"
          style="flex-basis: 60%;"
        >
          0
        </div>
        <div
          class="ant-splitter-bar"
        >
          <div
            class="ant-splitter-bar-bg"
          />
          <div
            class="ant-splitter-bar-area"
          />
          <div
            class="ant-splitter-bar-resizable"
          />
        </div>
        <div
          class="ant-splitter-panel"
          style="flex-basis: 40%;"
        >
          1
        </div>
      </div>
      ,
    </div>
    <div
      class="ant-splitter-bar"
    >
      <div
        class="ant-splitter-bar-bg"
      />
      <div
        class="ant-splitter-bar-area"
      />
      <div
        class="ant-splitter-bar-resizable"
      />
    </div>
    <div
      class="ant-splitter-panel"
      style="flex-basis: 50%;"
    >
      <div
        class="ant-splitter ant-splitter-vertical"
      >
        <div
          class="ant-splitter-panel"
          style="flex-basis: 70%;"
        >
          0
        </div>
        <div
          class="ant-splitter-bar"
        >
          <div
            class="ant-splitter-bar-bg"
          />
          <div
            class="ant-splitter-bar-area"
          />
          <div
            class="ant-splitter-bar-resizable"
          />
        </div>
        <div
          class="ant-splitter-panel"
          style="flex-basis: 30%;"
        >
          1
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter should correct render 1`] = `
<div>
  <div
    class="ant-splitter ant-splitter-horizontal"
  >
    <div
      class="ant-splitter-panel"
      style="flex-basis: 50%;"
    >
      0
    </div>
    <div
      class="ant-splitter-bar"
    >
      <div
        class="ant-splitter-bar-bg"
      />
      <div
        class="ant-splitter-bar-area"
      />
      <div
        class="ant-splitter-bar-resizable"
      />
    </div>
    <div
      class="ant-splitter-panel"
      style="flex-basis: 50%;"
    >
      1
    </div>
  </div>
</div>
`;
